### PR TITLE
Add Registry UI

### PR DESCRIPTION
**Pull request by CannonLock**
- Add Components for viewing Namespace
- Add forms for creating and updating Namespaces
- Update login page for CIlogin Users

### CODE REVIEW

**Comment by CannonLock** [comment]
@haoming29 This test failure isn't likely to be my error

**Comment by haoming29** [comment]
> @haoming29 This test failure isn't likely to be my error

Nope. That's a flaky test, will re-run it

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/page.tsx:118]: There seems to be no way to check the detail of a registration unless to edit it? Yes?

**Review by haoming29 — COMMENTED** [review]
We don't seem to allow no-authed users to view registry page? When I didn't login and go to `https://<registry-host>/` or `https://<registry-host>/view/registry/`, I will be redirected to the login page.

File: web_ui/frontend/components/Namespace.tsx
Comment (haoming29) [inline web_ui/frontend/components/Namespace.tsx:279]:
```suggestion
                        (authenticated?.role == "admin" || (authenticated?.user == namespace.admin_metadata.user_id && namespace.admin_metadata.status != "approved") ) &&
```

File: web_ui/frontend/components/Namespace.tsx
Comment (haoming29) [inline web_ui/frontend/components/Namespace.tsx:271]: When click the approve button, admin metadata will also be expanded. May be call `stopPropagation()` on those buttons?

File: web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx:115]: Can we also add a helper text here? "Pubkey is your origin's public JWKS. It must in JSON format, not JSON string".

If possible, I'd like to have a placeholder value to make it clear:
```json
{
    "keys": [
        {
            "alg": "ES256",
            "crv": "P-256",
            "kid": "<Your-kid>",
            "kty": "EC",
            "x": "",
            "y": ""
        }
    ]
}
```

File: web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx:164]: Also a help text here: "CILogon User Identifier of the user responsible for the security of the service"

File: web_ui/frontend/app/registry/namespace/register/page.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/namespace/register/page.tsx:82]: Can we redirect user to `/registry` page here after the alert? It's a bit weird to stay on an empty registration form once the user successfully submit their registration.

Comment (CannonLock) [reply web_ui/frontend/app/registry/namespace/register/page.tsx:82]: @haoming29 That was intentional incase they are looking to submit multiple entries at once. If you don't see that happening I will redirect.

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/page.tsx:71]: Can we add some text under "Pending Registration" or "Namespace Registry" clarifying who the namespace listed below belong to? When I login as a regular user it's a bit confusing to see "Pending Registration" and "Origin" and "Cache" as I don't know if this is others' namespace or mine.

The help text can be based on the user role:
* "Namespace Registry" 
  * Admin users: Namespace registry shows all registered namespaces in the federation. As an admin user, you can check pending namespace registration and review them. Approved registration are listed under "Origin" or "Cache" table.
  * Regular users: Namespace registry shows your registered namespaces in the federation. You can edit your registrations prior to approval. If you want to edit your registration that has been approved, please contact your registry administrator.
* "Pending Registration"
  * Admin users: "Pending Registration are namespaces waiting for review and approve or deny. If a cache is in pending status, the director will not use that cache"
  * Regular users: ""

Screenshot 2024-01-09 at 3 15 39 PM

Comment (haoming29) [reply web_ui/frontend/app/registry/page.tsx:71]: Also, I think currently if I logged in as a regular user, I won't be able to see others namespace, right? It seems that I can only see the one I owned. We still want to show a list of _all_ namespaces in the federation somewhere in the UI.

File: web_ui/frontend/app/registry/page.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/page.tsx:118]: If you click on the bar it will drop down with the namespace details.

File: web_ui/frontend/app/registry/page.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/page.tsx:117]: @haoming29 Why do you think that is? Could this be a API issue?

File: web_ui/frontend/components/Namespace.tsx
Comment (CannonLock) [reply web_ui/frontend/components/Namespace.tsx:271]: I do, I will look into why it isn't working.

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [reply web_ui/frontend/app/registry/page.tsx:118]: You are right. I though that's only admin_metadata but we don't need to see `pubkey` and `identity` so I guess that's all the details we want

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [reply web_ui/frontend/app/registry/page.tsx:71]: Ahh. Never mind. I think the regular user can still see _all_ namespace registrations (it was me who didn't register namespace under a different user and confused myself). Now the question is, how can a user tell if the namespace belongs to themselves?

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [inline web_ui/frontend/app/registry/page.tsx:77]: For admin users, they should be able to see a list of denied registrations. For regular users, it should be also helpful to see _their_ denied registrations. Right now denied registrations are gone.

File: web_ui/frontend/app/registry/page.tsx
Comment (haoming29) [reply web_ui/frontend/app/registry/page.tsx:117]: See suggested change in the comment below

File: web_ui/frontend/app/registry/page.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/page.tsx:77]: Agreed but no time this round to implement that.

File: web_ui/frontend/app/registry/page.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/page.tsx:71]: You can see all approved if you are not logged in.

File: web_ui/frontend/app/registry/page.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/page.tsx:117]: Ah I see you never intended user to be able to edit their approved namespaces. I did not realize.

File: web_ui/frontend/dev/data/pubkey.json
Comment (haoming29) [inline web_ui/frontend/dev/data/pubkey.json:1]: Why would we want to include this file here? If it's for example, we should remove the sensitive content like `kid` `x` and `y`

Comment (CannonLock) [reply web_ui/frontend/dev/data/pubkey.json:1]: These are public keys aren't they?

**Review by haoming29 — CHANGES_REQUESTED** [review]
Thanks for the PR! Left suggestions as comment.

File: web_ui/frontend/helpers/login.tsx
Comment (CannonLock) [reply web_ui/frontend/helpers/login.tsx:70]: Yeah, this token isn't really secure in any sense. The only way it could be abuse if we had a XSS vulnerability in which case they could do it even without it in sessionStorage.

File: web_ui/frontend/dev/data/pubkey.json
Comment (haoming29) [reply web_ui/frontend/dev/data/pubkey.json:1]: Yes, but also they are also specific to an origin (although the one here is probably from my local dev server). People can impersonate an origin with their public key and also "cybersquatting" namespace belonging to somebody else's origin, which could be a severe potential security issue.

File: web_ui/frontend/helpers/login.tsx
Comment (CannonLock) [reply web_ui/frontend/helpers/login.tsx:70]: The only point of the token is to confirm that the request is sent from our site. If they can steal it from sessionStorage they can send requests from our site as well so they could just get their own token and use it.

File: web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx
Comment (CannonLock) [reply web_ui/frontend/app/registry/namespace/components/NamespaceForm.tsx:164]: Fixed

File: web_ui/frontend/app/registry/namespace/register/page.tsx
Comment (jhiemstrawisc) [reply web_ui/frontend/app/registry/namespace/register/page.tsx:82]: My guess is no, unless people who manage a bunch of namespaces from topology decide to switch many of them over to Pelican.

Comment (haoming29) [reply web_ui/frontend/app/registry/namespace/register/page.tsx:82]: > My guess is no, unless people who manage a bunch of namespaces from topology decide to switch many of them over to Pelican.

In that case, I would recommend them to write their little python script to do it. It's going to be tedious and error prone to do it manually.

**Comment by haoming29** [comment]
LGTM after the revision! One last thing to add and we should be good to go.

> Can we add some text under "Pending Registration" and "Namespace Registry" clarifying who the namespace listed below belong to? When I login as a regular user it's a bit confusing to see "Pending Registration" and "Origin" and "Cache" as I don't know if this is others' namespace or mine.

> The help text can be based on the user role:
> * Under "Namespace Registry" 
>   * **Admin users**: Namespace registry shows all registered namespaces in the federation. As an admin user, you can check pending namespace registration and review them. Approved registration are listed under "Origin" or "Cache" table. Denied registrations are hidden.
>   * **Regular users**: Namespace registry shows all registered namespaces in the federation. You can edit your registrations prior to approval. If you want to edit your registration that has been approved, please contact your registry administrator.
> * Under "Pending Registration"
>   * **Admin users**: Pending Registration are namespaces waiting for your review and approve or deny. If a cache is in pending status, the director will not use that cache
>   * **Regular users**: Pending Registration are namespaces waiting for admin to review. Caches with pending status will not be used in the federation until they are approved

**Comment by CannonLock** [comment]
@haoming29 I made some tweaks in my changes to reduce the amount of text. It is hard to get people to read text so I tried to add some ui elements to answer your questions and your text suggestions where UI couldn't answer them all. 

Sorry about missing the issue you had to fix, I thought I saw a second suggested change for the input form but I looked through a couple times and couldn't find it.

**Review by haoming29 — APPROVED** [review]
LGTM with the revision. Thanks for putting this together.

**Comment by haoming29** [comment]
Looks like the only test failure is the client side test that's supposed by fixed by @joereuss12 soon? Will merge ours first.

**Pull request merged**
